### PR TITLE
[CALCITE-4936] Generalize FilterCalcMergeRule and ProjectCalcMergeRule

### PR DESCRIPTION
**Pull request by mgramin**
[CALCITE-4936] Generalize FilterCalcMergeRule and ProjectCalcMergeRule. Like CalcMergeRule, use only Project, Filter and Calc interfaces, which will expand the range of their applying.

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/rel/rules/ProjectCalcMergeRule.java [hunk 1]
```diff
@@ -82,8 +82,6 @@ public ProjectCalcMergeRule(RelBuilderFactory relBuilderFactory) {
             project.getRowType(),
             cluster.getRexBuilder());
     if (RexOver.containsOver(program)) {
-      LogicalCalc projectAsCalc = LogicalCalc.create(calc, program);
```
Comment (rubenada) [inline core/src/main/java/org/apache/calcite/rel/rules/ProjectCalcMergeRule.java:85]: Isn't this a breaking change? Currently the Project is removed, with this proposal the Project remains.

Comment (mgramin) [reply core/src/main/java/org/apache/calcite/rel/rules/ProjectCalcMergeRule.java:85]: Unfortunately, yes. But i assume that special rule `ProjectToCalcRule` can perform simple conversion from `LogicalProject` to `LogicalCalc`. Whereas `ProjectCalcMergeRule` will be able to stay more clear and more generalize.

Comment (zabetak) [reply core/src/main/java/org/apache/calcite/rel/rules/ProjectCalcMergeRule.java:85]: This is a small breaking change. I doubt people will notice but let's put a small note in `history.md` to be on the safe side. Moreover, please remove the now obsolete comments in lines 73-75 above.

**Review by zabetak — APPROVED** [review]
LGTM, only minor comments left.

General comment independent of this change: Do we really need `FilterCalcMergeRule` and `ProjectCalcMergeRule`? They look a bit redundant in the presence of `FilterToCalcRule` `ProjectToCalcRule` and `CalcMergeRule`.

**Comment by mgramin** [comment]
> General comment independent of this change: Do we really need `FilterCalcMergeRule` and `ProjectCalcMergeRule`? They look a bit redundant in the presence of `FilterToCalcRule` `ProjectToCalcRule` and `CalcMergeRule`.

`FilterCalcMergeRule` and `ProjectCalcMergeRule` produce only `LogicalCalc` also, it can make further optimizations  more difficult.

**Review by hsyuan — APPROVED** [review]
LGTM. Shall we merge it?

**Comment by rubenada** [comment]
@hsyuan ok for me, but please be aware that we are currently in code freeze since we are in the middle of building a new release. We could merge this patch once master re-opens.

**Comment by zabetak** [comment]
I will merge this in a few minutes.

**Pull request closed without merging**